File: 0016-3sum-closest/0016-3sum-closest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& n, int target) {
       sort(n.begin(),n.end());
        int ans = 0;
        int prev = INT_MAX;
        int i = 0;
        while(i<n.size() - 2){
            int j = i + 1;
            int k = n.size() - 1;
            while(j<k){
                int sum = n[i] + n[j] + n[k];
                int d = abs(target - sum);
                if(d<prev){
                    prev = d;
                    ans = sum;
                
                }
                if(sum < target){
                    j++;
                    } else{
                    k--;
                }
            }
            i++;
        }
        
        return ans;
    }
};

File: 0016-3sum-closest/0016-3sum-closest.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums);
        int n = nums.length;
        int res = nums[0] + nums[1] + nums[2];

        for(int i=0;i<n-2;i++) {
            int l = i+1, r = n-1;
            while(l<r) {
                int s = nums[i] + nums[l] + nums[r];
                if(Math.abs(target-s) < Math.abs(target-res)) {
                    res = s;
                }

                if(s==target) {
                    return s;
                } else if(s < target) {
                    l++;
                } else {
                    r--;
                }
            }
        }
        return res;
    }
}